### PR TITLE
Templates: Use IndentWithInclude

### LINKED ISSUE #115: Scriban templates: Use IndentWithInclude to simplify indentation
Use IndentWithInclude to simplify indentation. With this we can perhaps get rid of the whitespace parameter to indent code.

Scriban in its current version 3.0.3. still has a bug, which blocks this: https://github.com/lunet-io/scriban/issues/266

Check if the consistent use of `~` helps in creating well formated documents (see: https://github.com/lunet-io/scriban/blob/master/doc/language.md#14-whitespace-control)

### PR DESCRIPTION
**Pull request by badcel**
Fixes #115

### CODE REVIEW

File: Generator/Templates/GObject/class.sbntxt [hunk 1]
```diff
@@ -38,65 +38,58 @@ namespace {{ namespace }}
...
         protected internal {{ name }}(params ConstructParameter[] properties) : base(properties) {}
         
         #endregion
         
-        {{- end }}
-{{- end }}
-
+        {{~ end ~}}
+{{~ end ~}}
         #region Methods
```
Comment (na2axl) [inline Generator/Templates/GObject/class.sbntxt:71]: I think it's better to rename this region to `Native`

File: Generator/Templates/GObject/class.sbntxt [hunk 2]
```diff
@@ -38,65 +38,58 @@ namespace {{ namespace }}
...
-            
-            {{~for $method in all_methods
-                include 'method.sbntxt' $method dll_import '            ' name
-            end ~}}
+            {{~ for $method in all_methods }}
+            {{ include 'method.sbntxt' $method dll_import name ~}}
+            {{ end ~}}
             
             #endregion
         }
```
Comment (na2axl) [inline Generator/Templates/GObject/class.sbntxt:95]: Need a line break here before the `#endregion`

File: Generator/Templates/GObject/class.sbntxt [hunk 3]
```diff
@@ -38,65 +38,58 @@ namespace {{ namespace }}
...
+        {{~ end ~}}
+{{~ end ~}}
         #region Methods
         
         protected static{{' new'|get_if_parent}} class Native
         {
-        {{~if !fundamental && (string.size get_type_function) > 0 
-            get_type_function | write_dll_import dll_import '            '
-        ~}}
+
```
Comment (na2axl) [inline Generator/Templates/GObject/class.sbntxt:75]: Is this empty line intended ?

File: Generator/Templates/GObject/class.sbntxt [hunk 4]
```diff
@@ -38,65 +38,58 @@ namespace {{ namespace }}
...
-{{if !fundamental}}
+{{~ if !fundamental ~}}
         private static readonly TypeDescriptor GTypeDescriptor = TypeDescriptor.For("{{ type_name }}", {{ name }}.Native.get_type);
 
         #region Signals
         
-        {{- if signals_have_event_args }}
-        
+        {{~ if signals_have_event_args ~}}
         ///<summary>
```
Comment (na2axl) [inline Generator/Templates/GObject/class.sbntxt:47]: There is no space between the triple dash and the `summary` tag.

File: Generator/Templates/GObject/signal.sbntxt [hunk 5]
```diff
@@ -31,52 +30,48 @@ else
...
-            #region Properties
-            {{~ 
-            index = 0
-            for $param in $signal.parameters.parameters 
-                index = index + 1
-                $signal | write_summary ($indent + '    ')
-            ~}}
-            public {{ $param | write_type }} {{ $param.name | fix_identifier | make_pascal_case }} => Args[{{index}}].Extract<{{ $param | write_type }}>();
+{{~ if $signal.needs_signal_args ~}}
+///<summary>
```
Comment (na2axl) [inline Generator/Templates/GObject/signal.sbntxt:36]: There is no space between the triple slash and the `summary` tag.

File: Generator/Templates/GObject/signal.sbntxt [hunk 6]
```diff
@@ -31,52 +30,48 @@ else
...
+/// Indexer to connect {{ $signal_descriptor_name }} with an SignalHandler<{{$signal_generic_args}}>
```
Comment (na2axl) [inline Generator/Templates/GObject/signal.sbntxt:37]: I think we can use the `see` tag here to reference signal/event and signal handler in docs (and in intellisense).

File: Generator/Templates/GObject/signal.sbntxt [hunk 7]
```diff
@@ -31,52 +30,48 @@ else
...
-        ~}}
-        public event SignalHandler<{{ $signal_generic_args }}> {{ $signal_name }}
-        {
-            add => {{ $signal_descriptor_name }}.Connect(this, value);
-            remove => {{ $signal_descriptor_name }}.Disconnect(this, value);
-        }
-        
-        #endregion
+/// <summary>
+/// Signal (Event) Arguments for {{$signal_name}}
```
Comment (na2axl) [inline Generator/Templates/GObject/signal.sbntxt:45]: I think we can use the see tag here to reference the signal/event in docs (and in intellisense).

File: Generator/Templates/GObject/signal.sbntxt [hunk 8]
```diff
@@ -31,52 +30,48 @@ else
...
+/// </summary>
+public sealed class {{$signal_type}} : SignalArgs
+{            
+    #region Properties
+    {{~
```
Comment (na2axl) [inline Generator/Templates/GObject/signal.sbntxt:50]: Missing line break here

File: Generator/Templates/GObject/class.sbntxt [hunk 9]
```diff
@@ -38,65 +38,58 @@ namespace {{ namespace }}
...
     {
-{{if !fundamental}}
+{{~ if !fundamental ~}}
         private static readonly TypeDescriptor GTypeDescriptor = TypeDescriptor.For("{{ type_name }}", {{ name }}.Native.get_type);
 
         #region Signals
         
-        {{- if signals_have_event_args }}
-        
+        {{~ if signals_have_event_args ~}}
```
Comment (na2axl) [inline Generator/Templates/GObject/class.sbntxt:46]: An empty line is generated when this condition is false.

File: Generator/Templates/GObject/signal.sbntxt [hunk 10]
```diff
@@ -31,52 +30,48 @@ else
...
+/// Signal (Event) Arguments for {{$signal_name}}
+/// </summary>
+public sealed class {{$signal_type}} : SignalArgs
+{            
+    #region Properties
+    {{~ 
+    index = 0
+    for $param in $signal.parameters.parameters 
+        index = index + 1
+        '    ' + ($signal | get_summary)
```
Comment (na2axl) [inline Generator/Templates/GObject/signal.sbntxt:54]: One indentation level (4 spaces) is missing in the generated code here.

Comment (na2axl) [inline Generator/Templates/GObject/signal.sbntxt:54]: One indentation level (4 spaces) is missing in the generated code at this place

File: Generator/Templates/Shared/delegate.sbntxt [hunk 11]
```diff
@@ -3,9 +3,6 @@ using System.Runtime.InteropServices;
 
 namespace {{ namespace }}
 {
-{{
-include 'helper.sbntxt'
-this | write_summary '    '
-~}}
-    public delegate {{ return_value | resolve_type }} {{ name | fix_identifier }}({{ this | write_parameters }});
+    {{ include 'header.sbntxt' this ~}}
+    public delegate {{ return_value | resolve_type }} {{ name | fix_identifier }}({{ this | get_parameters }});
```
Comment (na2axl) [inline Generator/Templates/Shared/delegate.sbntxt:7]: The delegate get one more indentation level (4 spaces) in the generated code

File: Generator/Templates/Shared/enum.sbntxt [hunk 12]
```diff
@@ -2,13 +2,14 @@ using System;
...
+    {{ include 'header.sbntxt' this ~}}
     {{~ if has_flags ~}}
     [Flags]
     {{~ end ~}}
     public enum {{ name }} : long
     {
-{{~ for member in members
-include 'enum_member.sbntxt' member '        '
-end ~}}
+        {{~ for member in members }}
```
Comment (na2axl) [inline Generator/Templates/Shared/enum.sbntxt:11]: This generate an empty line

File: Generator/Templates/Shared/enum.sbntxt [hunk 13]
```diff
@@ -2,13 +2,14 @@ using System;
...
+        {{ include 'enum_member.sbntxt' member ~}}
+        {{ end }}
```
Comment (na2axl) [inline Generator/Templates/Shared/enum.sbntxt:13]: This generate an empty line

File: Generator/Templates/Shared/struct.sbntxt [hunk 14]
```diff
@@ -7,28 +7,25 @@ namespace {{ namespace }}
 {
     // AUTOGENERATED FILE - DO NOT MODIFY
 
-{{
-include 'helper.sbntxt'
-this | write_summary '    '
-this | write_obsolete_attribute '    '
-~}}
+    {{ include 'header.sbntxt' this ~}}
     [StructLayout(LayoutKind.Sequential)]
```
Comment (na2axl) [inline Generator/Templates/Shared/struct.sbntxt:11]: This line get one more indentation level (4 spaces) in the generated code

File: Generator/Templates/Shared/struct.sbntxt [hunk 15]
```diff
@@ -7,28 +7,25 @@ namespace {{ namespace }}
...
-end ~}}
+        {{~ for field in fields }}
+        {{ include 'field.sbntxt' field ~}}
+        {{ end ~}}
 
         #endregion
 
         #region Methods
 
         internal static class Native
```
Comment (na2axl) [inline Generator/Templates/Shared/struct.sbntxt:23]: We should rename this region `Native`

**Review by na2axl — APPROVED** [review]

**Comment by na2axl** [comment]
These updates are good for me, I've just left these little comments, but they are all negligible since the generated code works as expected :+1:

**Review by na2axl — APPROVED** [review]
Just some negligible things we can fix.

File: Generator/Templates/GObject/class.sbntxt (hunk 3 above)
Comment (badcel) [reply Generator/Templates/GObject/class.sbntxt:75]: Yes it provides the newline either infront of the `#region` or the `get_type` method.

File: Generator/Templates/GObject/class.sbntxt (hunk 9 above)
Comment (badcel) [reply Generator/Templates/GObject/class.sbntxt:46]: Looks good for me?

```C#
public partial class ParamSpec 
{
    #region Native
        
    protected static class Native
```

Do we have a problem with line breaks between windows / linux?

Comment (na2axl) [reply Generator/Templates/GObject/class.sbntxt:46]: The problem is under the `#region Signals` line, when there is no signal using `EventArgs` the output is like:

```cs
#region Signals


#region Signal1

...

#endregion

#endregion
```

File: Generator/Templates/Shared/delegate.sbntxt (hunk 11 above)
Comment (badcel) [reply Generator/Templates/Shared/delegate.sbntxt:7]: This looks good for me? Do you have an example?

File: Generator/Templates/Shared/enum.sbntxt (hunk 12 above)
Comment (badcel) [reply Generator/Templates/Shared/enum.sbntxt:11]: For me enums look fine, e.g.:

```C#
[Flags]
public enum ConnectFlags : long
{

    /// <summary>
    /// ...
    /// </summary>
    After = 1,
    
    /// <summary>
    /// ...
    /// </summary>
    Swapped = 2,
    
}        
```

File: Generator/Templates/Shared/struct.sbntxt (hunk 14 above)
Comment (badcel) [reply Generator/Templates/Shared/struct.sbntxt:11]: This works for me e.g.:

```C#
/// </summary>
[StructLayout(LayoutKind.Sequential)]
public partial struct FlagsClass
{
```

File: Generator/Templates/GObject/class.sbntxt (hunk 1 above)
Comment (badcel) [reply Generator/Templates/GObject/class.sbntxt:71]: Done

File: Generator/Templates/GObject/class.sbntxt (hunk 4 above)
Comment (badcel) [reply Generator/Templates/GObject/class.sbntxt:47]: Done

File: Generator/Templates/GObject/signal.sbntxt (hunk 5 above)
Comment (badcel) [reply Generator/Templates/GObject/signal.sbntxt:36]: Done

File: Generator/Templates/GObject/signal.sbntxt (hunk 6 above)
Comment (badcel) [reply Generator/Templates/GObject/signal.sbntxt:37]: I can't verify this. Intellisense behaves strange. I would probably postpone this until we have a generated documentation.

File: Generator/Templates/GObject/signal.sbntxt (hunk 7 above)
Comment (badcel) [reply Generator/Templates/GObject/signal.sbntxt:45]: I can't verify this. Intellisense behaves strange. I would probably postpone this until we have a generated documentation.

File: Generator/Templates/GObject/signal.sbntxt (hunk 8 above)
Comment (badcel) [reply Generator/Templates/GObject/signal.sbntxt:50]: Done

File: Generator/Templates/GObject/signal.sbntxt (hunk 10 above)
Comment (badcel) [reply Generator/Templates/GObject/signal.sbntxt:54]: Done

File: Generator/Templates/Shared/struct.sbntxt (hunk 15 above)
Comment (badcel) [reply Generator/Templates/Shared/struct.sbntxt:23]: Done

File: Generator/Templates/GObject/class.sbntxt (hunk 9 above)
Comment (badcel) [reply Generator/Templates/GObject/class.sbntxt:45]: Done

File: Generator/Templates/GObject/class.sbntxt (hunk 2 above)
Comment (badcel) [reply Generator/Templates/GObject/class.sbntxt:95]: For me it is generated correctly? E.g.:

```C#
            [DllImport("libgtk-3.so.0", EntryPoint = "gtk_button_new_with_mnemonic")]
            public static extern IntPtr new_with_mnemonic(string label);
                        
            #endregion
```

Comment (badcel) [reply Generator/Templates/GObject/class.sbntxt:95]: Done

File: Generator/Templates/Shared/delegate.sbntxt (hunk 11 above)
Comment (na2axl) [reply Generator/Templates/Shared/delegate.sbntxt:7]: @badcel I've discovered that this issue appear only when the delegate doesn't have a doc comment (for example some delegates in Gtk3)

File: Generator/Templates/Shared/enum.sbntxt (hunk 12 above)
Comment (na2axl) [reply Generator/Templates/Shared/enum.sbntxt:11]: The empty lines I mean are after and before opening and closing braces of the enum:

```cs
[Flags]
public enum ConnectFlags : long
{
    // EMPTY LINE HERE
    /// <summary>
    /// ...
    /// </summary>
    After = 1,
    
    /// <summary>
    /// ...
    /// </summary>
    Swapped = 2,
    // EMPTY LINE HERE
}        
```

But this completely negligible 🙂

File: Generator/Templates/Shared/struct.sbntxt (hunk 14 above)
Comment (na2axl) [reply Generator/Templates/Shared/struct.sbntxt:11]: This issue also appear only when the struct doesn't have a doc comment (like the AboutDialogClass struct)

File: Generator/Templates/Shared/enum.sbntxt (hunk 12 above)
Comment (badcel) [reply Generator/Templates/Shared/enum.sbntxt:11]: I think this is okay :smile:

**Comment by badcel** [comment]
I fixed the indentation offset. This results in an additional empty line if no doc comments are present which is preferable to a wrong indentation from my point of view. Unfortunately I don't see an easy way to fix this. Any ideas?

**Pull request merged**